README.md
This extension is used to provide description to ACPI predefine names (methods and PNPID). Mouse hover will triger the hints.
![demo](https://dev.azure.com/WilliamSuper/87a3c72c-732f-4c9e-96bf-a02f700f8615/_apis/git/repositories/bbe36879-d9cc-43e1-b8b9-87c624d2e566/items?path=%2FDemo.png&versionDescriptor%5BversionOptions%5D=0&versionDescriptor%5BversionType%5D=0&versionDescriptor%5Bversion%5D=master&resolveLfs=true&%24format=octetStream&api-version=5.0)

This is based on ACPI Spec version 6.4, https://uefi.org/specs/ACPI/6.4/, and the language definiation is based on CPP.

currently, the auto-complete is not implemented yet.

William Wu



CHANGELOG.md
# Change Log

All notable changes to the "acpihelper" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.
## [1.0]
- Add external key words support, and user can add Acpi keywords description in json file, the file is to be placed at C:\AcpiHelper\AcpiHelper.json

## [0.9]

- Add legacy PNPID support, which is based on http://download.microsoft.com/download/5/7/7/577a5684-8a83-43ae-9272-ff260a9c20e2/pnp_legacy.doc, find the link info from https://uefi.org/acpi 


## [0.8]

- Initial release

ToolsToGetAcpiPreDefinedNames/GetPreDefinedPnpId.py
import requests
from bs4 import BeautifulSoup
from bs4.diagnose import diagnose

class PreDefinedPnpId:
  PnpHtmlTxt = ""
  def __init__(self):
    try:
       with open('PnpId.Html','r') as TxtFo:
          TempPnpHtmlTxt = TxtFo.read()
       TxtFo.close()
       #print(TempPnpHtmlTxt)
       
       soup = BeautifulSoup(TempPnpHtmlTxt, 'lxml')
       if (soup.select("#acpi-device-ids") != None):
         self.PnpHtmlTxt = str(soup.find(['tbody']))
    except:
       print ("File read error with PnpId HTML")
    '''
    try:
       with open('PnpId.Html','r') as TxtFo:
          self.PnpHtmlTxt = TxtFo.read()
       TxtFo.close()
       soup = BeautifulSoup(self.PnpHtmlTxt, 'html.parser')
       if (soup.find("#acpi-device-ids") != None):
         self.PnpHtmlTxt = soup.find("#acpi-device-ids").string
       print(self.PnpHtmlTxt)
    except:
       print ("File read error with PnpId HTML")
    '''
    
  
  def Test(self):    
      soup = BeautifulSoup(self.PnpHtmlTxt, 'lxml')
      AllTdItem = soup.find_all(['p'])
      IndexValue = 0
      for child in AllTdItem:
        IndexValue = IndexValue + 1
        if ((IndexValue & 0x01) == 0x01): 
            print( " ")
            print(child.text + "  : ") 
        else: 
            print(child.text )

  def PrintPnpId(self):
    try:
      OutStr = ""
      soup = BeautifulSoup(self.PnpHtmlTxt, 'lxml')
      AllTdItem = soup.find_all(['p'])
      for IndexVal in range(len(AllTdItem)):
        if ((IndexVal & 0x01) == 0x00): 
          OutStr = OutStr + '\"' + AllTdItem[IndexVal].text + '\", ' 
          if ((IndexVal & 0x07) == 0x00):
             OutStr = OutStr + '\n'
      print (OutStr)
      return OutStr
    except:
      print ("Failed to parse pnp id")
      return None        
  ''' 
  Get Pnp Device ID
   - class started with .
   - id started with #
  '''
  def GetDeviceIdDesc(self, DescStr:str):
    try:
      
      soup = BeautifulSoup(self.PnpHtmlTxt, 'lxml')
      AllTdItem = soup.find_all(['p'])
      for IndexVal in range(len(AllTdItem)):
        if ((IndexVal & 0x01) == 0x00): 
            if (AllTdItem[IndexVal].text == DescStr):
               #print (AllTdItem[IndexVal + 1].text)
               return AllTdItem[IndexVal + 1].text
    except:
      print ("Failed to parse pnp id")
      return None
 
mPnpId = PreDefinedPnpId()
#mPnpId.GetDeviceIdDesc('PNP0C0F')
mPnpId.PrintPnpId()



ToolsToGetAcpiPreDefinedNames/GetPreDefinedNames.py
import requests
from bs4 import BeautifulSoup
from bs4.diagnose import diagnose

class PreDefinedPnpId:
  PnpHtmlTxt = ""
  def __init__(self):
    try:
       with open('PnpId.Html','r') as TxtFo:
          TempPnpHtmlTxt = TxtFo.read()
       TxtFo.close()
       
       soup = BeautifulSoup(TempPnpHtmlTxt, 'lxml')
       if (soup.select("#acpi-device-ids") != None):
         self.PnpHtmlTxt = str(soup.find(['tbody']))
    except:
       print ("File read error with PnpId HTML")
    
  def Test(self):    
      soup = BeautifulSoup(self.PnpHtmlTxt, 'lxml')
      AllTdItem = soup.find_all(['p'])
      IndexValue = 0
      for child in AllTdItem:
        IndexValue = IndexValue + 1
        if ((IndexValue & 0x01) == 0x01): 
            print( " ")
            print(child.text + "  : ") 
        else: 
            print(child.text )

  def PrintPnpId(self):
    try:
      OutStr = ""
      soup = BeautifulSoup(self.PnpHtmlTxt, 'lxml')
      AllTdItem = soup.find_all(['p'])
      for IndexVal in range(len(AllTdItem)):
        if ((IndexVal & 0x01) == 0x00): 
          if ((IndexVal > 0) and ((IndexVal & 0x03) == 0x00)): 
            OutStr = OutStr + '\n'
          OutStr = OutStr + '\"' + AllTdItem[IndexVal].text + '\", ' 
      print (OutStr)
      return OutStr
    except:
      print ("Failed to parse pnp id")
      return None        
  
  def PrintPnpId2(self):
    try:
      OutStr = ""
      soup = BeautifulSoup(self.PnpHtmlTxt, 'lxml')
      AllTdItem = soup.find_all(['p'])
      for IndexVal in range(len(AllTdItem)):
        if ((IndexVal & 0x01) == 0x00): 
          OutStr = OutStr + '|' + AllTdItem[IndexVal].text
      print (OutStr)
      return OutStr
    except:
      print ("Failed to parse pnp id")
      return None  
      
  def PrintPnpIdDesc(self):
    try:
      OutStr = ""
      soup = BeautifulSoup(self.PnpHtmlTxt, 'lxml')
      AllTdItem = soup.find_all(['p'])
      for IndexVal in range(len(AllTdItem)):
        if ((IndexVal & 0x01) == 0x01): 
          OutStr = OutStr + '\"' + AllTdItem[IndexVal].text + '\", \n' 
      print (OutStr)
      return OutStr
    except:
      print ("Failed to parse pnp id")
      return None        
  ''' 
  Get Pnp Device ID
   - class started with .
   - id started with #
  '''
  def GetDeviceIdDesc(self, DescStr:str):
    try:
      
      soup = BeautifulSoup(self.PnpHtmlTxt, 'lxml')
      AllTdItem = soup.find_all(['p'])
      for IndexVal in range(len(AllTdItem)):
        if ((IndexVal & 0x01) == 0x00): 
            if (AllTdItem[IndexVal].text == DescStr):
               #print (AllTdItem[IndexVal + 1].text)
               return AllTdItem[IndexVal + 1].text
    except:
      print ("Failed to parse pnp id")
      return None
      
class PreDefinedName:
  PreNameHtmlTxt = ""
  def __init__(self):
    try:
       with open('PredefinedNames.html','r') as TxtFo:
          PreNameHtmlTxt = TxtFo.read()
       TxtFo.close()
       #print(PreNameHtmlTxt)
       
       soup = BeautifulSoup(PreNameHtmlTxt, 'lxml')
       if (soup.select("#predefined-acpi-names") != None):
         self.PreNameHtmlTxt = str(soup.find(['tbody']))
    except:
       print ("File read error with Predefine Name HTML")

  def PrintPreNames(self):
    try:
      OutStr = ""
      soup = BeautifulSoup(self.PreNameHtmlTxt, 'lxml')
      AllTdItem = soup.find_all(['p'])
      for IndexVal in range(len(AllTdItem)):
        if ((IndexVal & 0x01) == 0x00): 
          if ((IndexVal > 0) and ((IndexVal & 0x07) == 0x00)): 
            OutStr = OutStr + '\n'
          OutStr = OutStr + '\"' + AllTdItem[IndexVal].text + '\", ' 
      print(OutStr)
      return OutStr
    except:
      return None 
  def PrintPreNames2(self):
    try:
      OutStr = ""
      soup = BeautifulSoup(self.PreNameHtmlTxt, 'lxml')
      AllTdItem = soup.find_all(['p'])
      for IndexVal in range(len(AllTdItem)):
        if ((IndexVal & 0x01) == 0x00): 
          OutStr = OutStr + '|' + AllTdItem[IndexVal].text  
      print(OutStr)
      return OutStr
    except:
      return None 
      
  def PrintPreNamesDec(self):
    try:
      OutStr = ""
      soup = BeautifulSoup(self.PreNameHtmlTxt, 'lxml')
      AllTdItem = soup.find_all(['p'])
      for IndexVal in range(len(AllTdItem)):
        if ((IndexVal & 0x01) == 0x01): 
          OutStr = OutStr + '\"' + AllTdItem[IndexVal].text + '\", \n' 
      print(OutStr)
      return OutStr
    except:
      return None   
  ''' 
  Get predefine names
  - class started with .
  - id started with #
  '''
  def GetPreDefineNameDesc(self, DescStr:str):
    try:
      soup = BeautifulSoup(self.PreNameHtmlTxt, 'lxml')
      AllTdItem = soup.find_all(['p'])
      for IndexVal in range(len(AllTdItem)):
        if ((IndexVal & 0x01) == 0x00): 
            if (AllTdItem[IndexVal].text == DescStr):
               #print (AllTdItem[IndexVal + 1].text)
               return AllTdItem[IndexVal + 1].text
    except:
      print ("Failed to parse predefine name")
      return None
 
mPreName = PreDefinedName()
mPrePnpName = PreDefinedPnpId()
print('let PreDefineName: string = [')
mPreName.PrintPreNames()
mPrePnpName.PrintPnpId()
print('];\n');

print('let PreDefNameHelpStr: string = [')
mPreName.PrintPreNamesDec()
mPrePnpName.PrintPnpIdDesc()
print('];\n');

mPreName.PrintPreNames2()
mPrePnpName.PrintPnpId2()





